# AML 3303 — Week 3 In-Class Lab (30 minutes)
### pandas + NumPy only

## Scenario

**NorthLine Electronics** is an online store. The ops team exported two tables for you:
`customers` and `orders`. They want a quick read on which product categories actually
make money, and which city orders the most.

The export is realistic, which means it is slightly messy: a missing quantity, a missing
price, a duplicate order, and one order from a customer who is not in the customer list.

Fill in every cell marked `# TODO`. The hints sit directly above each one.

## 0. Setup

Run this cell once. Do not change it.

In [2]:
# Run this cell first. It creates the two DataFrames you will work with.
import pandas as pd
import numpy as np

customers = pd.DataFrame({
    "customer_id": [1, 2, 3, 4, 5, 6, 7, 8],
    "name": ["Aarav Shah","Bianca Rossi","Chen Wei","Diya Patel",
             "Emeka Obi","Farah Haddad","Grace Kim","Hugo Martin"],
    "city": ["Ottawa","Toronto","Ottawa","Montreal",
             "Toronto","Calgary","Montreal","Ottawa"],
})

orders = pd.DataFrame({
    "order_id":   [101,102,103,104,105,106,107,108,109,110,
                   111,112,113,114,115,116,117,118,119,120,118,121],
    "customer_id":[  1,  2,  1,  3,  4,  5,  2,  6,  3,  7,
                     8,  5,  1,  4,  6,  7,  2,  8,  3,  5,  8,  9],
    "order_date": ["2024-05-02","2024-05-02","2024-05-04","2024-05-05","2024-05-07",
                   "2024-05-08","2024-05-09","2024-05-11","2024-05-12","2024-05-13",
                   "2024-05-15","2024-05-16","2024-05-18","2024-05-19","2024-05-20",
                   "2024-05-22","2024-05-23","2024-05-25","2024-05-26","2024-05-28",
                   "2024-05-25","2024-05-29"],
    "category":   ["Laptops","Headphones","Headphones","Monitors","Laptops",
                   "Keyboards","Monitors","Laptops","Keyboards","Headphones",
                   "Monitors","Laptops","Keyboards","Headphones","Monitors",
                   "Laptops","Keyboards","Headphones","Laptops","Monitors",
                   "Headphones","Keyboards"],
    "quantity":   [1,2,1,2,1,3,1,1,np.nan,1,4,1,2,1,1,1,1,2,1,2,2,1],
    "unit_price": [1299.99,89.50,89.50,219.00,1499.00,45.00,219.00,1099.00,45.00,np.nan,
                   219.00,1299.99,45.00,89.50,219.00,1499.00,45.00,89.50,1299.99,219.00,
                   89.50,45.00],
    "status":     ["delivered","delivered","returned","delivered","cancelled",
                   "delivered","delivered","delivered","delivered","delivered",
                   "delivered","returned","delivered","delivered","delivered",
                   "delivered","cancelled","delivered","delivered","delivered",
                   "delivered","delivered"],
})

print("orders:", orders.shape, " customers:", customers.shape)
orders.head()

orders: (22, 7)  customers: (8, 3)


,order_id,customer_id,order_date,category,quantity,unit_price,status
0,101,1,2024-05-02,Laptops,1.0,1299.99,delivered
1,102,2,2024-05-02,Headphones,2.0,89.50,delivered
2,103,1,2024-05-04,Headphones,1.0,89.50,returned
3,104,3,2024-05-05,Monitors,2.0,219.00,delivered
4,105,4,2024-05-07,Laptops,1.0,1499.00,cancelled


## 1. First look (3 min)

**Hints**
- `df.shape` → rows and columns
- `df.info()` → column types and non-null counts
- `df.describe()` → quick numeric summary
- `df["category"].value_counts()` → how many orders per category

In [10]:
import pandas as pd
import numpy as np

orders = pd.DataFrame({
    "order_id":   [101,102,103,104,105,106,107,108,109,110,
                   111,112,113,114,115,116,117,118,119,120,118,121],
    "customer_id":[  1,  2,  1,  3,  4,  5,  2,  6,  3,  7,
                     8,  5,  1,  4,  6,  7,  2,  8,  3,  5,  8,  9],
    "order_date": ["2024-05-02","2024-05-02","2024-05-04","2024-05-05","2024-05-07",
                   "2024-05-08","2024-05-09","2024-05-11","2024-05-12","2024-05-13",
                   "2024-05-15","2024-05-16","2024-05-18","2024-05-19","2024-05-20",
                   "2024-05-22","2024-05-23","2024-05-25","2024-05-26","2024-05-28",
                   "2024-05-25","2024-05-29"],
    "category":   ["Laptops","Headphones","Headphones","Monitors","Laptops",
                   "Keyboards","Monitors","Laptops","Keyboards","Headphones",
                   "Monitors","Laptops","Keyboards","Headphones","Monitors",
                   "Laptops","Keyboards","Headphones","Laptops","Monitors",
                   "Headphones","Keyboards"],
    "quantity":   [1,2,1,2,1,3,1,1,np.nan,1,4,1,2,1,1,1,1,2,1,2,2,1],
    "unit_price": [1299.99,89.50,89.50,219.00,1499.00,45.00,219.00,1099.00,45.00,np.nan,
                   219.00,1299.99,45.00,89.50,219.00,1499.00,45.00,89.50,1299.99,219.00,
                   89.50,45.00],
    "status":     ["delivered","delivered","returned","delivered","cancelled",
                   "delivered","delivered","delivered","delivered","delivered",
                   "delivered","returned","delivered","delivered","delivered",
                   "delivered","cancelled","delivered","delivered","delivered",
                   "delivered","delivered"],
})

print(orders.shape)
orders.info()
print("\nOrders per category:")
print(orders["category"].value_counts())

(22, 7)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   order_id     22 non-null     int64  
 1   customer_id  22 non-null     int64  
 2   order_date   22 non-null     object 
 3   category     22 non-null     object 
 4   quantity     21 non-null     float64
 5   unit_price   21 non-null     float64
 6   status       22 non-null     object 
dtypes: float64(2), int64(2), object(3)
memory usage: 1.3+ KB

Orders per category:
category
Laptops       6
Headphones    6
Monitors      5
Keyboards     5
Name: count, dtype: int64


## 2. A first (wrong) answer (5 min)

Compute revenue per category **before** cleaning anything, counting delivered orders only.
Keep this result: you will compare against it at the end.

**Hints**
- Filter rows: `orders[orders["status"] == "delivered"]`
- New column: `df["revenue"] = df["quantity"] * df["unit_price"]`
- Group: `df.groupby("category")["revenue"].sum().sort_values(ascending=False)`

In [11]:
delivered_orders = orders[orders["status"] == "delivered"].copy()
delivered_orders["revenue"] = delivered_orders["quantity"] * delivered_orders["unit_price"]
revenue_raw = delivered_orders.groupby("category")["revenue"].sum().sort_values(ascending=False)
revenue_raw

,revenue
category,
Laptops,5197.98
Monitors,2190.00
Headphones,626.50
Keyboards,270.00


## 3. Find the problems (5 min)

**Hints**
- Duplicates: `df.duplicated().sum()`, and `df[df.duplicated(keep=False)]` to see them
- Missing values: `df.isna().sum()`
- Orphan rows: `~orders["customer_id"].isin(customers["customer_id"])`

In [12]:
# 1. how many fully duplicated rows are there? Show them.
print("Number of fully duplicated rows:", orders.duplicated().sum())
print("Fully duplicated rows:")
print(orders[orders.duplicated(keep=False)])

# 2. how many missing values in each column?
print("\nMissing values per column:")
print(orders.isna().sum())

# 3. which orders belong to a customer who is NOT in the customers table?
print("\nOrders from unknown customers:")
print(orders[~orders["customer_id"].isin(customers["customer_id"])])

Number of fully duplicated rows: 1
Fully duplicated rows:
    order_id  customer_id  order_date    category  quantity  unit_price  \
17       118            8  2024-05-25  Headphones       2.0        89.5   
20       118            8  2024-05-25  Headphones       2.0        89.5   

       status  
17  delivered  
20  delivered  

Missing values per column:
order_id       0
customer_id    0
order_date     0
category       0
quantity       1
unit_price     1
status         0
dtype: int64

Orders from unknown customers:
    order_id  customer_id  order_date   category  quantity  unit_price  \
21       121            9  2024-05-29  Keyboards       1.0        45.0   

       status  
21  delivered  


## 4. Clean it (7 min)

**Hints**
- `orders_clean = orders.drop_duplicates()`
- Fill a price with the median for that category:
  `df.groupby("category")["unit_price"].transform("median")`
- `df["quantity"].fillna(1)` — one unit is a reasonable default for this shop
- `pd.to_datetime(df["order_date"])`
- Make quantity an integer at the end: `.astype(int)`

In [14]:
orders_clean = orders.copy()

# 1. drop duplicate rows
orders_clean.drop_duplicates(inplace=True)

# 2. fill missing unit_price with the median price of the same category
orders_clean["unit_price"] = orders_clean.groupby("category")["unit_price"].transform(lambda x: x.fillna(x.median()))

# 3. fill missing quantity with 1, then convert the column to int
orders_clean["quantity"] = orders_clean["quantity"].fillna(1)
orders_clean["quantity"] = orders_clean["quantity"].astype(int)

# 4. convert order_date to datetime
orders_clean["order_date"] = pd.to_datetime(orders_clean["order_date"])

orders_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21 entries, 0 to 21
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   order_id     21 non-null     int64         
 1   customer_id  21 non-null     int64         
 2   order_date   21 non-null     datetime64[ns]
 3   category     21 non-null     object        
 4   quantity     21 non-null     int64         
 5   unit_price   21 non-null     float64       
 6   status       21 non-null     object        
dtypes: datetime64[ns](1), float64(1), int64(3), object(2)
memory usage: 1.3+ KB


## 5. NumPy: revenue column (3 min)

Cancelled and returned orders earn **no** revenue.

**Hints**
- `np.where(condition, value_if_true, value_if_false)`
- Condition: `orders_clean["status"] == "delivered"`
- `np.round(x, 2)` if you want tidy numbers

In [15]:
orders_clean["revenue"] = np.where(
    (orders_clean["status"] == "delivered"),
    orders_clean["quantity"] * orders_clean["unit_price"],
    0
)

print(f"Total Revenue: ${orders_clean['revenue'].sum():.2f}")
print(f"Mean Order Value: ${orders_clean['revenue'].mean():.2f}")
print(f"Largest Single Order: ${orders_clean['revenue'].max():.2f}")

Total Revenue: $8239.98
Mean Order Value: $392.38
Largest Single Order: $1499.00


## 6. Merge and answer the business question (5 min)

**Hints**
- `orders_clean.merge(customers, on="customer_id", how="left")`
- A left merge keeps the order from the unknown customer, with `city` as NaN. That is
  usually what you want: do not silently delete money.
- `.groupby("city")["revenue"].sum().sort_values(ascending=False)`

In [16]:
merged_df = orders_clean.merge(customers, on="customer_id", how="left")
revenue_by_city = merged_df.groupby("city")["revenue"].sum().sort_values(ascending=False)
revenue_by_city

,revenue
city,
Ottawa,4227.98
Montreal,1678.00
Calgary,1318.00
Toronto,971.00


## 7. Compare with step 2 (2 min)

**Hint:** `pd.DataFrame({"raw": revenue_raw, "clean": revenue_clean})` lines the two up side by side.

In [17]:
revenue_clean = orders_clean[orders_clean["status"] == "delivered"].groupby("category")["revenue"].sum().sort_values(ascending=False)
comparison = pd.DataFrame({"raw": revenue_raw, "clean": revenue_clean})
comparison

,raw,clean
category,,
Laptops,5197.98,5197.98
Monitors,2190.00,2190.00
Headphones,626.50,537.00
Keyboards,270.00,315.00


**Question:** two categories changed. Which ones, and what caused each change?

*Your answer:*

Two categories changed: **Headphones** and **Keyboards**.

- **Headphones:** The raw revenue was 626.50 dollars and the cleaned revenue was 537.00 dollars. The decrease is due to two factors:
  1.  A fully duplicated order (order ID 118) for Headphones, which contributed 179.00 dollars in revenue, was dropped.
  2.  A missing `unit_price` for order ID 110 (Headphones) was filled with the median price of 89.50 dollars, adding 89.50 dollars in revenue.
  The net effect was a **net decrease** of $89.50 (calculated as a decrease of $179.00 from the dropped duplicate, and an increase of $89.50 from the filled missing price, resulting in $179.00 - $89.50 = $89.50 net decrease).

- **Keyboards:** The raw revenue was 270.00 dollars and the cleaned revenue was 315.00 dollars. The increase is due to a missing `quantity` for order ID 109 (Keyboards) being filled with 1. This added $45.00 (1 * $45.00) to the revenue for this category.

## 8. Wrap up (optional, if time)

Write two sentences for the NorthLine ops team: which category to push, and one data
quality issue they should fix at the source.

*Your answer:*

NorthLine Electronics should focus on promoting **Laptops**, as they consistently generate the highest revenue both before and after data cleaning. To improve data quality, the ops team should implement a system to prevent duplicate order entries at the source, as these can lead to inflated revenue figures and inaccurate reporting.

---